sw/device/lib/crypto/api.h

```diff
@@ -0,0 +1,1378 @@
+ * a `ciphertext` data as output. The partial data bytes are stored
+ * back in the `gcm_context` partial_data parameter and combined with
+ * subsequent bytes or padded to block length for the last block. The
+ * intermediate result is stored in the `gcm_context` state parameter.
+ *
+ * #aes_gcm_init and if needed #aes_gcm_aad_update should be called
+ * before calling this API.
+ * @param gcm_context Pointer to the GCM context struct
+ * @param plaintext Input data to be encrypted and authenticated
+ * @param ciphertext Encrypted output data, same length as input data
+ * @return crypto_status_t The result of the init operation
+ */
+crypto_status_t aes_gcm_encrypt_update(aes_gcm_context_t *gcm_context,
+                                       const crypto_uint8_buf_t plaintext,
+                                       crypto_uint8_buf_t *ciphertext);
+
+/**
+ * Performs the AES-GCM authenticated encryption final operation.
+ *
+ * This function pads the remaining partial data and encrypts the
+ * block. Computes GHASH of the updated state after and stores the
+ * result back in `gcm_context` state parameter, for tag generation.
+ *
+ * #aes_gcm_encrypt_update should be called before calling this API.
+ *
+ * @param gcm_context Pointer to the GCM context struct
+ * @param ciphertext Encrypted output data, same length as input data
+ * @return crypto_status_t The result of the init operation
+ */
+crypto_status_t aes_gcm_encrypt_final(aes_gcm_context_t *gcm_context,
+                                      crypto_uint8_buf_t *ciphertext);
+
+/**
+ * Performs the authentication tag generation operation.
+ *
+ * Performs GCTR on the final state value stored in the `gcm_context`
+ * and returns the generated `auth_tag`.
+ *
+ * #aes_gcm_encrypt_final should be called before using this API.
+ *
+ * @param gcm_context Pointer to the GCM context struct
+ * @param auth_tag Authentication tag
+ * @return crypto_status_t The result of the tag generate operation
+ */
+crypto_status_t aes_gcm_generate_tag(aes_gcm_context_t *gcm_context,
+                                     crypto_uint8_buf_t *auth_tag);
+
+/**
+ * Performs the pre-step for AES-GCM tag verification operation.
+ *
+ * Note: Tag should be verified before decrypting ciphertext.
+ *
+ * This function computes the final state value from the `ciphertext`
+ * and AAD data. The partial data bytes are stored back in the
+ * `gcm_context` partial_data parameter. The intermediate result is
+ * stored in the `gcm_context` state parameter.
+ *
+ * #aes_gcm_init and aes_gcm_aad_update should be called before using
+ * this API.
+ *
+ * @param gcm_context Pointer to the GCM context struct
+ * @param ciphertext Encrypted input data for tag generation
+ * @return crypto_status_t The return status of the verify update operation
+ */
+crypto_status_t aes_gcm_verify_tag_update(aes_gcm_context_t *gcm_context,
+                                          const crypto_uint8_buf_t ciphertext);
```

Comment:
This doesn't affect the pointer inside, unfortunately. I think we need a `crypto_const_uint8_buf_t`...